File: frontendmasters/rust-for-typescript-devs/typescript/src/submarine.ts
import { readFileSync } from "fs";
import { join } from "path";

function getInput(): string {
  return readFileSync(join(__dirname, "..", "..", "input.txt")).toString();
}

function parseLine(line: string): [number, number] {
  const [dir, a] = line.split(" ");
  const amount = +a;

  if (dir === "forward") {
    return [amount, 0];
  } else if (dir === "up") {
    return [0, -amount];
  }
  return [0, amount];
}

const out = getInput()
  .split("\n")
  .map((x) => parseLine(x))
  .reduce(
    (acc, amount) => {
      acc[0] += amount[0];
      acc[1] += amount[1];
      return acc;
    },
    [0, 0]
  );

console.log(out, out[0] * out[1]);

File: frontendmasters/rust-for-typescript-devs/rust/src/bin/submarine.rs
fn get_input(file_path: &str) -> String {
    return std::fs::read_to_string(file_path).expect("Should have been able to read the file");
}

#[derive(Debug)]
struct Point {
    x: i32,
    y: i32,
}

fn parse_line(line: &str) -> Point {
    let (dir, amount) = line.split_once(" ").expect("must contain a whitespace");
    let amount = str::parse::<i32>(amount).expect("second arg must be an int");

    if dir == "forward" {
        return Point { x: amount, y: 0 };
    } else if dir == "up" {
        return Point { x: 0, y: -amount };
    }
    return Point { x: 0, y: amount };
}

fn main() {
    let result = get_input("../../../input.txt").lines()
    .map(parse_line)
    .fold(Point {x: 0, y: 0}, |mut acc, point| {
        acc.x += point.x;
        acc.y += point.y;
        return acc;
    });

    println!("{:?}, {}", result, result.x*result.y)
}
